### PR TITLE
Allow merging of extra section in a deep way.

### PR DESCRIPTION
**Pull request by LionsAd**
### Motivation

I am trying to use composer-merge-plugin with a composer-patches variation, which is late enough for composer-merge-plugin to have merged all sections.

My root composer.json has something like:

```json
{
    "extra": {
        "merge-plugin": {
            "merge-extra": true,
            "merge-deep": true,
            "include": "composer.local.json"
        },
        "patches": {
            "wikimedia/composer-merge-plugin": {
                "Add tests for merge-deep option": "patches/add-tests-for-merge-deep-option.diff"
            },
            "anothervendor/some-project": {
                "some-patch": "patches/some-patch.diff"
            }
        }
    }
}
```

while my composer.local.json has:

```
{
    "extra": {
        "patches": {
            "wikimedia/composer-merge-plugin": {
                "Allow merging of sections in a deep way": "patches/add-merge-deep-option.diff"
            },
            "somevendor/some-project": {
                "some-patch": "patches/some-patch.diff"
            }
        }
    }
}
```

### Approach

This adds a merge-deep option, which only really is meaningful for the extra sections using an utility function from Drupal.

This makes the merge plugin much more useful for plugins that have deeper structures.

This also allows both kinda workflows that the merge plugin is used with:

* Defining some base skeleton files and including that, then overriding some keys (replace = TRUE)
* Overriding the global composer.json with a local variant.

Hopefully no-one will need this together, because that use case is not yet 'done', however I feel this is complex enough as is, to be a follow-up discussion.

### Notes

Unfortunately Drupal 8 at the moment does not allow to reference the NestedArray component directly via a subtree split, so I had to copy the code for now.

- Also see:

* https://github.com/composer/composer/issues/5232 - where I am suggesting to give composer-merge-plugin a chance to always be run after init.
* https://github.com/douggreen/drupal-composer-installer/pull/9 - where I am using this functionality.

### CODE REVIEW

**Comment by LionsAd** [comment]
Tests pass on everything except PHP 5.3.3, so I guess it is travis not my code that fails the tests.

**Comment by LionsAd** [comment]
Squashed and rebased.

**Comment by LionsAd** [comment]
Fixed tests.

**Comment by LionsAd** [comment]
Fixed tests and test coverage to be back to 100%.

**Comment by LionsAd** [comment]
@bd808 This is ready for review now :).

**Comment by LionsAd** [comment]
@bd808 Could I have some feedback on this PR, please?

**Comment by bd808** [comment]
Since this feature flag only affects merging of the `extra` section I think the naming of the flag and the methods that implement it should be less generic. Maybe `merge-extra-deep`?

I think I'd rather see `ExtraPackage::arrayMerge` made static and renamed something like `ExtraPackage::mergeExtraArray`.  `ExtraPackage::arrayMergeDeep` should introduced as a method in a static utility class. That will make it easier to both document the origin of the code and to possibly replace it at a later date when and if we can import a small library from Drupal.

**Comment by LionsAd** [comment]
@bd808 Done!

**Comment by LionsAd** [comment]
@bd808 ping, this is ready since a while.

File: src/NestedArray.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * This software may be modified and distributed under the terms of the MIT
+ * license. See the LICENSE file for details.
+ */
+
+namespace Wikimedia\Composer;
+
+/**
+ * Helper class to merge arrays in a deep fashion.
+ *
+ * @author Bryan Davis <bd808@bd808.com>
```
Comment (bd808) [inline src/NestedArray.php]: I didn't write this one. :)

**Comment by bd808** [comment]
Rebased, squashed, and merged with a few minor cosmetic fix-ups as 6e95cde

**Pull request closed without merging**